File: buffer/gpu/src/jvmMain/java/com/wsr/gpu/JShape.java
package com.wsr.gpu;

class JShape {
    public native void transposeD2(
            long x,
            int xi,
            int xj,
            long result,
            long runtime
    );

    public native void transposeD3(
            long x,
            int xi,
            int xj,
            int xk,
            int axisI,
            int axisJ,
            int axisK,
            long result,
            long runtime
    );

    public native void transposeD4(
            long x,
            int xi,
            int xj,
            int xk,
            int xl,
            int axisI,
            int axisJ,
            int axisK,
            int axisL,
            long result,
            long runtime
    );

    public native void sliceD1(
            long x,
            int start,
            int end,
            int step,
            long result,
            long runtime
    );

    public native void sliceD2(
            long x,
            int xi,
            int xj,
            int axis,
            int start,
            int end,
            int step,
            long result,
            long runtime
    );

    public native void sliceD3(
            long x,
            int xi,
            int xj,
            int xk,
            int axis,
            int start,
            int end,
            int step,
            long result,
            long runtime
    );
}

File: buffer/cpu/src/jvmAndroidMain/kotlin/com/wsr/knist/cpu/shape/JShape.kt
package com.wsr.knist.cpu.shape


object JShape {
    external fun transposeD2(x: Long, xi: Int, xj: Int, result: Long)

    external fun transposeD3(
        x: Long,
        xi: Int,
        xj: Int,
        xk: Int,
        axisI: Int,
        axisJ: Int,
        axisK: Int,
        result: Long,
    )

    external fun transposeD4(
        x: Long,
        xi: Int,
        xj: Int,
        xk: Int,
        xl: Int,
        axisI: Int,
        axisJ: Int,
        axisK: Int,
        axisL: Int,
        result: Long,
    )

    external fun transposeD5(
        x: Long,
        xi: Int,
        xj: Int,
        xk: Int,
        xl: Int,
        xm: Int,
        axisI: Int,
        axisJ: Int,
        axisK: Int,
        axisL: Int,
        axisM: Int,
        result: Long,
    )

    external fun sliceD1(x: Long, start: Int, end: Int, step: Int, result: Long)

    external fun sliceD2(
        x: Long,
        xi: Int,
        xj: Int,
        axis: Int,
        start: Int,
        end: Int,
        step: Int,
        result: Long,
    )

    external fun sliceD3(
        x: Long,
        xi: Int,
        xj: Int,
        xk: Int,
        axis: Int,
        start: Int,
        end: Int,
        step: Int,
        result: Long,
    )

    external fun copyIntoD1(x: Long, result: Long, start: Int, end: Int, step: Int)

    external fun copyIntoD2(
        x: Long,
        result: Long,
        ri: Int,
        rj: Int,
        axis: Int,
        start: Int,
        end: Int,
        step: Int,
    )

    external fun copyIntoD3(
        x: Long,
        result: Long,
        ri: Int,
        rj: Int,
        rk: Int,
        axis: Int,
        start: Int,
        end: Int,
        step: Int,
    )

    external fun unfoldD1(
        x: Long,
        xi: Int,
        xj: Int,
        b: Int,
        window: Int,
        stride: Int,
        dilation: Int,
        padding: Int,
        result: Long,
    )

    external fun unfoldD2(
        x: Long,
        xi: Int,
        xj: Int,
        xk: Int,
        b: Int,
        window: Int,
        stride: Int,
        dilation: Int,
        padding: Int,
        result: Long,
    )

    external fun foldD1(
        x: Long,
        xi: Int,
        xj: Int,
        xk: Int,
        b: Int,
        stride: Int,
        dilation: Int,
        padding: Int,
        result: Long,
    )

    external fun foldD2(
        x: Long,
        xi: Int,
        xj: Int,
        xk: Int,
        xl: Int,
        b: Int,
        stride: Int,
        dilation: Int,
        padding: Int,
        result: Long,
    )

    external fun flipD3(x: Long, xi: Int, xj: Int, xk: Int, axis: Int, result: Long)
}
